### PR TITLE
Add JSInterpreter tests for breakpoints and switch statements

### PR DESCRIPTION
**Pull request by joshlory**

### CODE REVIEW

File: apps/src/codegen.js [hunk 1]
```diff
@@ -559,19 +559,17 @@ exports.isNextStepSafeWhileUnwinding = function (interpreter) {
...
-  var cnt = session.getLength();
-  var cuml = 0, nlLength = session.getDocument().getNewLineCharacter().length;
-  cumulativeLength.push(cuml);
-  var text = session.getLines(0, cnt);
-  for (var i = 0; i < cnt; i++) {
-    cuml += text[i].length + nlLength;
-    cumulativeLength.push(cuml);
-  }
-  return cumulativeLength;
+exports.calculateCumulativeLength = function (code) {
```
Comment (islemaster) [inline apps/src/codegen.js:632]: Now that this is a pure function, would you add a few tests that document its expected behavior and edge cases?

**Comment by islemaster** [comment]
:microphone: :man: :dancers: :fire: :fire: :fire_engine:
Dude, you're a rockstar.  Nice work.

File: apps/src/codegen.js (hunk 1 above)
Comment (joshlory) [inline apps/src/codegen.js:632]: Good call, added in 79680cd4c49ae7adaa3fe35c3d86ff0fbddbacb7.

File: apps/src/turtle/turtle.js [hunk 2]
```diff
@@ -697,13 +697,12 @@ Artist.prototype.evalCode = function(code) {
...
+  var userCode = this.studioApp_.editor.getValue();
   this.code = dropletUtils.generateCodeAliases(dropletConfig, 'Turtle');
   this.userCodeStartOffset = this.code.length;
-  this.code += this.studioApp_.editor.getValue();
+  this.code += userCode;
   this.userCodeLength = this.code.length - this.userCodeStartOffset;
-
-  var session = this.studioApp_.editor.aceEditor.getSession();
-  this.cumulativeLength = codegen.aceCalculateCumulativeLength(session);
+  this.cumulativeLength = codegen.calculateCumulativeLength(userCode);
```
Comment (cpirich) [inline apps/src/turtle/turtle.js]: these changes are not needed if you approve https://github.com/code-dot-org/code-dot-org/pull/6849 first

File: apps/src/JSInterpreter.js [hunk 3]
```diff
@@ -649,9 +653,6 @@ JSInterpreter.prototype.selectCurrentCode = function (highlightClass) {
  * of the userCode area, the return value is -1
  */
 JSInterpreter.prototype.getUserCodeLine = function () {
-  if (this.studioApp.hideSource) {
-    return -1;
-  }
```
Comment (cpirich) [inline apps/src/JSInterpreter.js:713]: My concern with deleting this is that we've made the shared apps (`hideSource === true`) slower. The good news is they probably should have been running slower for cases where we weren't running at maxSpeed. But I wonder if you can provide a fast exit for the `maxSpeed` case? Things running too slowly in the interpreter has been a recurring theme, and keeping the "shared app" case running at the fastest possible speed seems like a good thing.

For your test cases below, you can probably ensure that you're not running at maxSpeed, true?

**Comment by cpirich** [comment]
Looks good. Just one concern related to performance for the `hideSource` scenario at max speed...

**Comment by joshlory** [comment]
@cpirich I think this is ready for another pass.  The core of the new behavior is here:
```javascript
// NOTE:
// (1) When running with no source visible AND at max speed, always set
//   `userCodeRow` to -1. We'll never hit a breakpoint and speed is max.
// (2) When running with no source visible or at max speed, call a simple
//   function to just get the line number. Need to check `inUserCode` to
//   maybe stop at a breakpoint, or add a `speed(n)` delay.
// (3) Otherwise call a function that also highlights the code.
var selectCodeFunc;
if (this.studioApp.hideSource && atMaxSpeed) {
  selectCodeFunc = function () { return -1; };
} else if (atMaxSpeed || this.studioApp.hideSource) {
  selectCodeFunc = this.getUserCodeLine;
} else {
  selectCodeFunc = this.selectCurrentCode;
}
```

**Comment by joshlory** [comment]
@cpirich I think this is ready for another pass. The core of the new behavior is here:

``` javascript
// NOTE:
// (1) When running with no source visible AND at max speed, always set
//   `userCodeRow` to -1. We'll never hit a breakpoint or need to add delay.
// (2) When running with no source visible OR at max speed, call a simple
//   function to just get the line number. Need to check `inUserCode` to
//   maybe stop at a breakpoint, or add a `speed(n)` delay.
// (3) Otherwise call a function that also highlights the code.
var selectCodeFunc;
if (this.studioApp.hideSource && atMaxSpeed) {
  selectCodeFunc = function () { return -1; };
} else if (this.studioApp.hideSource || atMaxSpeed) {
  selectCodeFunc = this.getUserCodeLine;
} else {
  selectCodeFunc = this.selectCurrentCode;
}
```

File: apps/src/JSInterpreter.js (hunk 3 above)
Comment (joshlory) [inline apps/src/JSInterpreter.js:713]: Fixed in ac9ed40d4c4596512f284c178ad409fe1aeec1fc.

**Pull request merged**